markets/storageadapter/ondealexpired.go

```diff
@@ -0,0 +1,152 @@
+package storageadapter
+
+import (
+	"context"
+
+	market0 "github.com/filecoin-project/specs-actors/actors/builtin/market"
+
+	"github.com/filecoin-project/lotus/build"
+	"github.com/filecoin-project/lotus/chain/actors/builtin/market"
+	"github.com/filecoin-project/lotus/chain/events"
+	"github.com/filecoin-project/lotus/chain/events/state"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/ipfs/go-cid"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-fil-markets/storagemarket"
+	"github.com/filecoin-project/go-state-types/abi"
+	sealing "github.com/filecoin-project/lotus/extern/storage-sealing"
+)
+
+type demEventsAPI interface {
+	Called(check events.CheckFunc, msgHnd events.MsgHandler, rev events.RevertHandler, confidence int, timeout abi.ChainEpoch, mf events.MsgMatchFunc) error
+	StateChanged(check events.CheckFunc, scHnd events.StateChangeHandler, rev events.RevertHandler, confidence int, timeout abi.ChainEpoch, mf events.StateMatchFunc) error
+}
+
+type demChainAPI interface {
+	ChainHead(context.Context) (*types.TipSet, error)
+}
+
+type DealExpiryManagerAPI interface {
+	demEventsAPI
+	demChainAPI
+	GetCurrentDealInfo(ctx context.Context, tok sealing.TipSetToken, proposal *market.DealProposal, publishCid cid.Cid) (sealing.CurrentDealInfo, error)
+}
+
+type dealExpiryManagerAdapter struct {
+	demEventsAPI
+	demChainAPI
+	*sealing.CurrentDealInfoManager
+}
+
+type DealExpiryManager struct {
+	demAPI    DealExpiryManagerAPI
+	dsMatcher *dealStateMatcher
+}
+
+func NewDealExpiryManager(ev demEventsAPI, ch demChainAPI, tskAPI sealing.CurrentDealInfoTskAPI, dsMatcher *dealStateMatcher) *DealExpiryManager {
+	dim := &sealing.CurrentDealInfoManager{
+		CDAPI: &sealing.CurrentDealInfoAPIAdapter{CurrentDealInfoTskAPI: tskAPI},
+	}
+
+	adapter := &dealExpiryManagerAdapter{
+		demEventsAPI:           ev,
+		demChainAPI:            ch,
+		CurrentDealInfoManager: dim,
+	}
+	return newDealExpiryManager(adapter, dsMatcher)
+}
+
+func newDealExpiryManager(demAPI DealExpiryManagerAPI, dsMatcher *dealStateMatcher) *DealExpiryManager {
+	return &DealExpiryManager{demAPI: demAPI, dsMatcher: dsMatcher}
+}
+
+func (mgr *DealExpiryManager) OnDealExpiredOrSlashed(ctx context.Context, publishCid cid.Cid, proposal market0.DealProposal, onDealExpired storagemarket.DealExpiredCallback, onDealSlashed storagemarket.DealSlashedCallback) error {
+	head, err := mgr.demAPI.ChainHead(ctx)
+	if err != nil {
+		return xerrors.Errorf("client: failed to get chain head: %w", err)
+	}
+
+	prop := market.DealProposal(proposal)
+	res, err := mgr.demAPI.GetCurrentDealInfo(ctx, head.Key().Bytes(), &prop, publishCid)
```

Comment:
Technically we can reorg after getting this info

Reply:
We can, but in order to fix that we would need to change the way the state change API works.
I suggest we merge this improvement first and then tackle the state change API changes.